gsd/gsd.c

```diff
@@ -749,51 +679,872 @@ int gsd_read_header(struct gsd_handle* handle)
+        total_bytes_written += bytes_written;
+    }
+
+    // sync the expanded index
+    retval = fsync(handle->fd);
+    if (retval != 0)
+    {
+        free(buf);
+        return GSD_ERROR_IO;
+    }
+
+    // free the copy buffer
+    free(buf);
+
+    // update the header
+    handle->header.index_location = new_index_location;
+    handle->file_size = handle->header.index_location + total_bytes_written;
+    handle->header.index_allocated_entries = size_new;
+
+    // write the new header out
+    ssize_t bytes_written
+        = gsd_io_pwrite_retry(handle->fd, &(handle->header), sizeof(struct gsd_header), 0);
+    if (bytes_written != sizeof(struct gsd_header))
+    {
+        return GSD_ERROR_IO;
+    }
+
+    // sync the updated header
+    retval = fsync(handle->fd);
+    if (retval != 0)
+    {
+        return GSD_ERROR_IO;
+    }
+
+    // remap the file index
+    retval = gsd_index_buffer_map(&handle->file_index, handle);
+    if (retval != 0)
+    {
+        return retval;
+    }
+
+    return GSD_SUCCESS;
+}
+
+/** @internal
+    @brief Flush the write buffer.
+
+    gsd_write_frame() writes small data chunks into the write buffer. It adds index entries for
+    these chunks to gsd_handle::buffer_index with locations offset from the start of the write
+    buffer. gsd_flush_write_buffer() writes the buffer to the end of the file, moves the index
+    entries to gsd_handle::frame_index and updates the location to reference the beginning of the
+    file.
+
+    @param handle Handle to flush the write buffer.
+    @returns GSD_SUCCESS on success or GSD_* error codes on error
+*/
+static int gsd_flush_write_buffer(struct gsd_handle* handle)
+{
+    if (handle == NULL)
+    {
+        return GSD_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (handle->write_buffer.size == 0)
+    {
+        // nothing to do
+        return GSD_SUCCESS;
+    }
+
+    if (handle->buffer_index.size == 0)
+    {
+        // error: bytes in buffer, but no index for them
+        return GSD_ERROR_INVALID_ARGUMENT;
+    }
+
+    // write the buffer to the end of the file
+    uint64_t offset = handle->file_size;
+    ssize_t bytes_written = gsd_io_pwrite_retry(handle->fd,
+                                                handle->write_buffer.data,
+                                                handle->write_buffer.size,
+                                                offset);
+
+    if (bytes_written == -1 || bytes_written != handle->write_buffer.size)
+    {
+        return GSD_ERROR_IO;
+    }
+
+    handle->file_size += handle->write_buffer.size;
+
+    // reset write_buffer for new data
+    handle->write_buffer.size = 0;
+
+    // move buffer_index entries to file_index
+    size_t i;
+    for (i = 0; i < handle->buffer_index.size; i++)
+    {
+        struct gsd_index_entry* new_index;
+        int retval = gsd_index_buffer_add(&handle->frame_index, &new_index);
+        if (retval != GSD_SUCCESS)
+        {
+            return retval;
+        }
```

Comment:
Is it important to role back changes if an error is encountered, or is it fine to leave the buffer with partially written information?

Reply:
If an error occurs here, there will be some data written to the file that has no index entry pointing to it. Rolling back would only save a small amount of file size and not change the correctness of the file. Previously written frames will still be readable and the frame being written to when the error occurs will either have no data chunks or a subset of the data chunks (e.g. if the user hit a file space quota when part way through writing the index for the frame).

There is only so much we can do about consistency without judicious use of `fsync` and the corresponding loss in performance.